File: internal/runtime/py/templates/_cc_runtime/fs.py
"""File store backed by S3."""

from . import _client


def connect(id):
    """Return a client for the file store declared as ``persist_fs(id)``."""
    bucket = _client.env("CC_FS_%s_BUCKET" % _client.slug(id), "persist_fs", id)
    return Bucket(id, bucket, _client.client("s3"))


class Bucket:
    def __init__(self, id, bucket, s3):
        self.id = id
        self._bucket = bucket
        self._s3 = s3

    def _key(self, key):
        return str(key).lstrip("/")

    def read(self, key):
        """Return the bytes stored at ``key``.

        Raises FileNotFoundError when absent, matching the local emulation
        rather than leaking botocore's ClientError to callers.
        """
        try:
            return self._s3.get_object(Bucket=self._bucket, Key=self._key(key))["Body"].read()
        except self._s3.exceptions.NoSuchKey:
            raise FileNotFoundError(key) from None

    def write(self, key, data):
        """Store ``data`` at ``key``."""
        if not isinstance(data, bytes):
            data = str(data).encode("utf-8")
        self._s3.put_object(Bucket=self._bucket, Key=self._key(key), Body=data)

    def delete(self, key):
        """Remove ``key`` if present."""
        self._s3.delete_object(Bucket=self._bucket, Key=self._key(key))

    def exists(self, key):
        """Whether ``key`` is present."""
        try:
            self._s3.head_object(Bucket=self._bucket, Key=self._key(key))
            return True
        except Exception:
            return False

    def list(self, prefix=""):
        """Every key under ``prefix``, sorted."""
        out = []
        token = None
        while True:
            kwargs = {"Bucket": self._bucket, "Prefix": prefix}
            if token:
                kwargs["ContinuationToken"] = token
            page = self._s3.list_objects_v2(**kwargs)
            out.extend(obj["Key"] for obj in page.get("Contents", []))
            if not page.get("IsTruncated"):
                return sorted(out)
            token = page.get("NextContinuationToken")

    def __repr__(self):
        return "<Bucket %r (s3)>" % self.id

File: internal/lang/node/templates/_cloudcc_runtime/fs.js
// File store backed by S3.

import { DeleteObjectCommand, GetObjectCommand, HeadObjectCommand, ListObjectsV2Command, PutObjectCommand, S3Client } from "@aws-sdk/client-s3";

import { common, env, slug } from "./client.js";

export function connect(id) {
  const bucket = env(`CLOUDCC_FS_${slug(id)}_BUCKET`, "persist", id);
  return new FileStore(id, bucket, new S3Client(common()));
}

export class FileStore {
  constructor(id, bucket, client) {
    this.id = id;
    this._bucket = bucket;
    this._client = client;
  }

  _key(key) {
    return String(key).replace(/^\/+/, "");
  }

  async read(key) {
    try {
      const out = await this._client.send(
        new GetObjectCommand({ Bucket: this._bucket, Key: this._key(key) }),
      );
      return Buffer.from(await out.Body.transformToByteArray());
    } catch (err) {
      // Match the SDK's local emulation, which throws rather than leaking the
      // provider's own error shape to callers.
      if (err?.name === "NoSuchKey" || err?.$metadata?.httpStatusCode === 404) {
        const e = new Error(`no such key: ${key}`);
        e.code = "ENOENT";
        throw e;
      }
      throw err;
    }
  }

  async write(key, data) {
    await this._client.send(
      new PutObjectCommand({ Bucket: this._bucket, Key: this._key(key), Body: data }),
    );
  }

  async delete(key) {
    await this._client.send(
      new DeleteObjectCommand({ Bucket: this._bucket, Key: this._key(key) }),
    );
  }

  async exists(key) {
    try {
      await this._client.send(
        new HeadObjectCommand({ Bucket: this._bucket, Key: this._key(key) }),
      );
      return true;
    } catch {
      return false;
    }
  }

  async list(prefix = "") {
    const out = [];
    let token;
    do {
      const page = await this._client.send(
        new ListObjectsV2Command({ Bucket: this._bucket, Prefix: prefix, ContinuationToken: token }),
      );
      for (const obj of page.Contents ?? []) {
        out.push(obj.Key);
      }
      token = page.IsTruncated ? page.NextContinuationToken : undefined;
    } while (token);
    return out.sort();
  }
}
